src/fu-engine-repair.c

```diff
@@ -0,0 +1,441 @@
+			    "Can't be reverted since kernel lockdown was disabled.");
+			return FALSE;
+		}
+
+		return grubby_set(FALSE, "lockdown=confidentiality", error);
+	}
+}
+
+static gboolean
+grubby_set_iommu(gboolean enable, GError **error)
+{
+	gboolean ret;
+	if (enable) {
+		return grubby_set(TRUE, "iommu=force", error);
+	} else {
+		ret = is_kernel_parameter("iommu=force", error);
+		if (!*error && ret) {
+			g_printf("OK iommu ret %d\n", ret);
+			return grubby_set(FALSE, "iommu=force", error);
+		} else {
+			g_set_error_literal(error,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_NOTHING_TO_DO,
+					    "iommu was not set.");
+			return FALSE;
+		}
+	}
+}
+
+static gboolean
+fu_engine_repair_kernel_lockdown(FuEngine *engine, const gchar *action, GError **error)
+{
+	FuSecurityAttrs *attrs;
+	FwupdSecurityAttr *attr;
+	guint flags;
+
+	attrs = fu_engine_get_host_security_attrs(engine);
+	if (!attrs)
+		return FALSE;
+
+	attr = fu_security_attrs_get_by_appstream_id(attrs, FWUPD_SECURITY_ATTR_ID_UEFI_SECUREBOOT);
+	if (!attr)
+		return FALSE;
+
+	flags = fwupd_security_attr_get_flags(attr);
+
+	if (!g_strcmp0(action, "undo")) {
+		if (flags == FWUPD_SECURITY_ATTR_FLAG_SUCCESS) {
+			g_set_error_literal(
+			    error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOTHING_TO_DO,
+			    "Kernel lockdown can't be disabled when secure boot is enabled.");
+			return FALSE;
+		} else {
+			return grubby_set_lockdown(FALSE, error);
+		}
+	}
+
+	return grubby_set_lockdown(TRUE, error);
+}
+
+static gboolean
+fu_engine_repair_iommu(const gchar *action, GError **error)
+{
+	if (!is_grubby_installed(error)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    FWUPD_ERROR_NOTHING_TO_DO,
+				    "Grubby was not installed.");
+		return FALSE;
+	}
+
+	if (!g_strcmp0(action, "undo"))
+		return grubby_set_iommu(FALSE, error);
+
+	if (is_kernel_parameter("iommu=", error) || is_kernel_parameter("intel_iommu=", error) ||
+	    is_kernel_parameter("amd_iommu=", error)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOTHING_TO_DO,
+				    "IOMMU had been already set.");
+		return FALSE;
+	}
+
+	return grubby_set_iommu(TRUE, error);
+}
+
+static gboolean
+fu_engine_repair_or_unsupport(FuEngine *engine,
+			      const gchar *appstream_id,
+			      const gchar *action,
+			      GError **error)
+{
+	FuSecurityAttrs *attrs;
+	FwupdSecurityAttr *attr;
+	GHashTable *settings = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, g_free);
+
+	attrs = fu_engine_get_host_security_attrs(engine);
+	if (!attrs)
+		return FALSE;
+
+	attr = fu_security_attrs_get_by_appstream_id(attrs, appstream_id);
+	if (!attr)
+		return FALSE;
+
+	/* test fixable BIOS settings */
+	if (fwupd_security_attr_get_bios_setting_id(attr) != NULL &&
+	    fwupd_security_attr_get_bios_setting_current_value(attr) != NULL &&
+	    fwupd_security_attr_get_bios_setting_target_value(attr) != NULL) {
+		/* TODO: should we revert the bios settings? */
```

Comment:
I think you should allow undoing them, specifically if they're not read only. Secure boot for example is read only once set. If any others are setting read only incorrectly we can do fixups in the BIOS settings creation.

Reply:
Thank you. If we have to revert the config, the previous setting should be kept somewhere, for example: creating a new variable called bios_setting_previous_value and updating the hsi_history table.